consensus/src/transaction_shuffler/fairness/mod.rs

```diff
@@ -0,0 +1,165 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::transaction_shuffler::{
+    fairness::{
+        conflict_key::ConflictKeyRegistry, conflict_zone::ConflictZone, pending_zone::PendingZone,
+    },
+    TransactionShuffler,
+};
+use aptos_types::transaction::SignedTransaction;
+use itertools::zip_eq;
+use selection_tracker::SelectionTracker;
+use std::collections::BTreeSet;
+
+pub(crate) mod conflict_key;
+mod conflict_zone;
+mod pending_zone;
+mod selection_tracker;
+
+#[cfg(test)]
+mod tests;
+
+type TxnIdx = usize;
+
+const NUM_CONFLICT_ZONES: usize = 3;
+
+struct FairnessShuffler {
+    sender_conflict_window_size: usize,
+    module_conflict_window_size: usize,
+    entry_fun_conflict_window_size: usize,
+}
+
+impl TransactionShuffler for FairnessShuffler {
+    fn shuffle(&self, txns: Vec<SignedTransaction>) -> Vec<SignedTransaction> {
+        let conflict_key_registries = ConflictKeyRegistry::build_registries(&txns);
+        let order = FairnessShufflerImpl::new(self, &conflict_key_registries).shuffle();
+        reorder(txns, order)
+    }
+}
+
+fn reorder(txns: Vec<SignedTransaction>, order: Vec<TxnIdx>) -> Vec<SignedTransaction> {
+    assert_eq!(txns.len(), order.len());
+    order.into_iter().map(|idx| txns[idx].clone()).collect()
```

Comment:
should we maybe assert here that we are not taking same transaction twice?

Reply:
That would be too heavy for an assertion?  Upcoming proptest will cover it.

Reply:
I meant have a bool vec, mark used fields as you go through, and assert we don't try to include the same one twice?

if you think halting the chain is too aggressive, we should log error!

Reply:
Too verbose an assertion to my taste but I can add it.

Reply:
oh... the equivalent is actually already asserted in `selection_tracker.select()`